File: GdPicture/WinForm/C#/PDF to PDF-OCR/Form1.cs
using System.Diagnostics;
using System.IO;
using System.Windows.Forms;
using GdPicture.OCR.Plugins;
using GdPicture14;



namespace PDF_TO_PDF_OCR
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
        }

        private readonly GdPicturePDF _nativePdf = new GdPicturePDF();
        private bool _cancellationPending;
        private Stopwatch _stopwatch = new Stopwatch();

        private void Form1_Load(object sender, System.EventArgs e)
        {
            LicenseManager oLicenseManager = new LicenseManager(); //Go to http://www.gdpicture.com/download-gdpicture/ to get a 1 month trial key unlocking all features of the toolkit.
            oLicenseManager.RegisterKEY("XXX"); //Please, replace XXX by a valid demo or commercial license key.

            txtDictsPath.Text = oLicenseManager.GetRedistPath() + "OCR\\";

            _nativePdf.OcrPagesProgress += this.OcrPagesProgress;
            _nativePdf.BeforePageOcr += this.BeforePageOcr;
            _nativePdf.OcrPagesDone += this.OcrPagesDone;
            _nativePdf.PageBitmapOcrReady += this.PageBitmapOcrReady;

            //GdPictureTesseract.AttachPlugin(_nativePdf); //Uncomment to use GdPicture OCR as external engine.
#if USE_OMNIPAGE_OCR
            OmniPage.AttachPlugin(_nativePdf);//Uncomment to use OmniPage OCR as external engine.
#endif
        }


        private void txtDictsPath_TextChanged(object sender, System.EventArgs e)
        {
            cmbLang.Items.Clear();
            if (Directory.Exists(txtDictsPath.Text))
            {
                DirectoryInfo d = new DirectoryInfo(txtDictsPath.Text);
                FileInfo[] files = d.GetFiles("*.traineddata");
                foreach (FileInfo file in files)
                {
                    cmbLang.Items.Add(Path.GetFileNameWithoutExtension(file.Name));
                }
            }
        }


        private void PageBitmapOcrReady(int PageNo, int ImageID)
        {
            //we're doing nothing yet.
        }



        private void BeforePageOcr(int pageNo, ref bool cancel)
        {
            //we're doing nothing yet.
        }


        private void OcrPagesProgress(GdPictureStatus status, int pageNo, int processed, int count, ref bool cancel)
        {
            ProgressBar1.Maximum = count;
            ProgressBar1.Value = processed;

            if (status != GdPictureStatus.OK)
            {
                if (!_cancellationPending)
                {
                    if (MessageBox.Show("An error occured on page " + pageNo + ". Do you want to cancel the process? Status: " + status, "error", MessageBoxButtons.YesNo, MessageBoxIcon.Stop) == DialogResult.Yes)
                    {
                        cancel = true;
                        _cancellationPending = true;
                    }
                }
            }
        }


        private void OcrPagesDone(GdPictureStatus status)
        {
            if (status == GdPictureStatus.OK)
            {
                _stopwatch.Stop();
                string outputFilePath = txtInputFile.Text.Substring(0, txtInputFile.Text.Length - 4) + "_ocr.pdf";

                status = chkIncSave.Checked ? _nativePdf.SaveToFileInc(outputFilePath) : _nativePdf.SaveToFile(outputFilePath, true);

                if (status != GdPictureStatus.OK)
                {
                    MessageBox.Show("Can't save file " + outputFilePath + " . Error: " + status);
                }
                else
                {
                    Process.Start(outputFilePath);
                }
                MessageBox.Show("Done! Please check file " + outputFilePath + "\r\nEllaped: " + _stopwatch.Elapsed.ToString());
            }
            else
            {
                MessageBox.Show("An error occured. Status: " + status);
            }

            _nativePdf.CloseDocument();
            UnlockUi();
        }


        private void LockUi()
        {
            this.Enabled = false;
            ProgressBar1.Visible = true;
            ProgressBar1.Value = 0;
        }


        private void UnlockUi()
        {
            this.Enabled = true;
            ProgressBar1.Visible = false;
        }


        private void btStartOCR_Click(object sender, System.EventArgs e)
        {
            string inputFilePath = txtInputFile.Text;

            if (!string.IsNullOrEmpty(inputFilePath))
            {
                if (!string.IsNullOrEmpty(cmbLang.Text))
                {
                    bool documentLoaded = false;
                    float resolution = float.Parse(txtDPI.Text);

                    if (GdPictureDocumentUtilities.GetDocumentFormat(inputFilePath) != GdPicture14.DocumentFormat.DocumentFormatPDF)
                    {
                        //converting input file to PDF.
                        using (GdPictureDocumentConverter gdpictureDocumentConverter = new GdPictureDocumentConverter())
                        {
                            if (gdpictureDocumentConverter.LoadFromFile(inputFilePath) == GdPictureStatus.OK)
                            {
                                string tmpFilePath = Path.GetTempFileName();
                                Stream stream = File.Create(tmpFilePath, 4096, FileOptions.DeleteOnClose);

                                if (gdpictureDocumentConverter.SaveAsPDF(stream, PdfConformance.PDF1_5) != GdPictureStatus.OK)
                                {
                                    MessageBox.Show("Can't convert file to PDF: " + inputFilePath);
                                    return;
                                }

                                _nativePdf.LoadFromStream(stream, true);
                                documentLoaded = true;
                            }
                            else
                            {
                                MessageBox.Show("Can't open input document: " + inputFilePath);
                                return;
                            }
                        }
                    }

                    if (documentLoaded || _nativePdf.LoadFromFile(inputFilePath, false) == GdPictureStatus.OK)
                    {
                        if (_nativePdf.IsEncrypted())
                        {   //PDF is encrypted, try to decrypt by using empty password
                            if (_nativePdf.SetPassword("") == GdPictureStatus.OK)
                            {
                                MessageBox.Show("This PDF is password protected", "operation cancelled", MessageBoxButtons.OK, MessageBoxIcon.Stop);
                                _nativePdf.CloseDocument();
                                return;
                            }
                        }

                        LockUi();
                        _cancellationPending = false;

                        int threadCount = chkMultithread.Checked ? 0 : 1;
                        OCRMode ocrMode = chkFavorSpeed.Checked ? OCRMode.FavorSpeed : OCRMode.FavorAccuracy;

                        _stopwatch.Restart(); //go go go!
                        _nativePdf.SetOcrPageOrientationDetection(chkEnablePageOrientationDetection.Checked);
                        _nativePdf.OcrPages("*", threadCount, cmbLang.Text, txtDictsPath.Text, "", resolution, ocrMode, 0, false);
                    }
                    else
                    {
                        MessageBox.Show("Can't open file: " + inputFilePath);
                    }
                }
                else
                {
                    MessageBox.Show("Please, select a language to recognize.");
                }
            }
            else
            {
                MessageBox.Show("Please, select a file to convert.");
            }
        }


        private void btBrowsInput_Click(object sender, System.EventArgs e)
        {
            using (OpenFileDialog openFileDialog1 = new OpenFileDialog())
            {
                openFileDialog1.FileName = "";
                openFileDialog1.Multiselect = false;

                if (openFileDialog1.ShowDialog() == DialogResult.OK)
                {
                    txtInputFile.Text = openFileDialog1.FileName;
                }
            }
        }


        private void btBrowsOCRDict_Click(object sender, System.EventArgs e)
        {
            using (FolderBrowserDialog folderBrowserDialog1 = new FolderBrowserDialog())
            {
                if (folderBrowserDialog1.ShowDialog() == DialogResult.OK)
                {
                    txtDictsPath.Text = folderBrowserDialog1.SelectedPath;
                }
            }
        }
    }
}

File: GdPicture/WinForm/vb.net/PDF to PDF-OCR/Form1.vb
Option Strict On
Option Explicit On

Imports System.IO
Imports GdPicture14

Public Class Form1

    Private ReadOnly _nativePdf As New GdPicturePDF
    Private _cancellationPending As Boolean

    Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
        Dim oLicenseManager As New GdPicture14.LicenseManager() 'Go to http://www.gdpicture.com/download-gdpicture/ to get a 1 month trial key unlocking all features of the toolkit.
        oLicenseManager.RegisterKEY("XXXX") 'Please, replace XXXX by a valid demo or commercial license key.

        txtDictsPath.Text = oLicenseManager.GetRedistPath() + "OCR\"

        AddHandler _nativePdf.OcrPagesProgress, AddressOf Me.OcrPagesProgress
        AddHandler _nativePdf.BeforePageOcr, AddressOf Me.BeforePageOcr
        AddHandler _nativePdf.OcrPagesDone, AddressOf Me.OcrPagesDone
    End Sub

    Private Sub txtDictsPath_TextChanged(sender As Object, e As EventArgs) Handles txtDictsPath.TextChanged
        If Directory.Exists(txtDictsPath.Text) Then
            Dim d As New DirectoryInfo(txtDictsPath.Text)
            Dim files As FileInfo() = d.GetFiles("*.traineddata")
            For Each file As FileInfo In files
                cmbLang.Items.Add(Path.GetFileNameWithoutExtension(file.Name))
            Next
        End If
    End Sub

    Private Sub BeforePageOcr(ByVal pageNo As Integer, ByRef cancel As Boolean)
        If _nativePdf.PageHasText(False) Then
            If MessageBox.Show("The page " + pageNo.ToString() + " of the pdf already has text. Do you want to skip it?", "The page has text", MessageBoxButtons.YesNo, MessageBoxIcon.Stop) = DialogResult.Yes Then
                cancel = True
            End If
        End If
    End Sub

    Private Sub OcrPagesProgress(ByVal status As GdPictureStatus, ByVal pageNo As Integer, ByVal processed As Integer, ByVal count As Integer, ByRef cancel As Boolean)
        ProgressBar1.Maximum = count
        ProgressBar1.Value = processed
        If status <> GdPictureStatus.OK Then
            If Not _cancellationPending Then
                If MessageBox.Show("An error occured on page " + pageNo.ToString() + ". Do you want to cancel the process? Status: " + status.ToString(), "error", MessageBoxButtons.YesNo, MessageBoxIcon.Stop) = DialogResult.Yes Then
                    cancel = True
                    _cancellationPending = True
                End If
            End If

        End If
    End Sub

    Private Sub OcrPagesDone(ByVal status As GdPictureStatus)
        If status = GdPictureStatus.OK Then
            Dim outputFilePath As String = txtInputFile.Text.Substring(0, txtInputFile.Text.Length - 4) + "_ocr.pdf"

            If chkIncSave.Checked Then
                status = _nativePdf.SaveToFileInc(outputFilePath)
            Else
                status = _nativePdf.SaveToFile(outputFilePath, True)
            End If

            If status <> GdPictureStatus.OK Then
                MessageBox.Show("Can't save file " + outputFilePath + " . Error: " + status.ToString())
            Else
                Process.Start(outputFilePath)
            End If
            MessageBox.Show("Done! Please check file " + outputFilePath)
        Else
            MessageBox.Show("An error occured. Status: " + status.ToString())
        End If

        _nativePdf.CloseDocument()
        UnlockUi()
    End Sub

    Private Sub LockUi()
        Me.Enabled = False
        ProgressBar1.Visible = True
        ProgressBar1.Value = 0
    End Sub

    Private Sub UnlockUi()
        Me.Enabled = True
        ProgressBar1.Visible = False
    End Sub

    Private Sub btStartOCR_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btStartOCR.Click
        If txtInputFile.Text <> "" Then
            If cmbLang.Text <> "" Then
                Dim resolution As Single = Single.Parse(txtDPI.Text)

                If _nativePdf.LoadFromFile(txtInputFile.Text, False) = GdPictureStatus.OK Then
                    If _nativePdf.IsEncrypted() Then
                        'PDF is encrypted, try to decrypt by using empty password
                        If Not _nativePdf.SetPassword("") <> GdPictureStatus.OK Then
                            MessageBox.Show("This PDF is password protected", "operation cancelled", MessageBoxButtons.OK, MessageBoxIcon.Stop)
                            _nativePdf.CloseDocument()
                            Return
                        End If
                    End If

                    LockUi()
                    _cancellationPending = False

                    Dim threadCount As Integer
                    If chkMultithread.Checked Then
                        threadCount = 0 '0 means automatic number of thread computation
                    Else
                        threadCount = 1
                    End If

                    _nativePdf.OcrPages("*", threadCount, cmbLang.Text, txtDictsPath.Text, "", resolution)
                Else
                    MessageBox.Show("Can't open file: " & txtInputFile.Text)
                End If
            Else
                MessageBox.Show("Please, select a language to recognize.")
            End If
        Else
            MessageBox.Show("Please, select a file to convert.")
        End If
    End Sub

    Private Sub btBrowsInput_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btBrowsInput.Click
        OpenFileDialog1.FileName = ""
        OpenFileDialog1.Filter = "PDF Document (*.pdf)|*.pdf"
        OpenFileDialog1.Multiselect = False
        If OpenFileDialog1.ShowDialog() = DialogResult.OK Then
            txtInputFile.Text = OpenFileDialog1.FileName
        End If
    End Sub

    Private Sub btBrowsOCRDict_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btBrowsOCRDict.Click
        If FolderBrowserDialog1.ShowDialog = DialogResult.OK Then
            txtDictsPath.Text = FolderBrowserDialog1.SelectedPath
        End If
    End Sub

End Class